### PR TITLE
[SW-2487] - Kubernetes API shutdown - SW servlets enabled

### PR DESCRIPTION
**Pull request by Pscheidl**
https://h2oai.atlassian.net/browse/SW-2487

### CODE REVIEW

**Comment by mn-mikke** [comment]
@Pscheidl Can we add a test suite which disables H2O-3 endpoints and perform a simple df conversions? Ideally. the suite would run under Spark master set to `local-cluster[2,1, 2048]`

- Some conversion tests: https://github.com/h2oai/sparkling-water/blob/b2f2d55009845a4be9e5d9f0fd3d9b63140addca/core/src/test/scala/ai/h2o/sparkling/backend/converters/DataFrameConverterTestSuite.scala

- Usage of `local-cluster` mode: https://github.com/h2oai/sparkling-water/blob/8e16ddb09fd50e2c97c943ea33bb4dde9907d2cd/core/src/test/scala/ai/h2o/sparkling/ConfigurationPropertiesTestSuite.scala

**Comment by Pscheidl** [comment]
Of course. Thanks.

**Comment by Pscheidl** [comment]
@mn-mikke The API only says the endpoint are always enabled under all circumstances, but there is no option to turn them off - that's not the point. Turning them off would mean not registering them at all.

A much more sensible test would be to test this on K8S  directly ?

**Comment by Pscheidl** [comment]
Attempted to add a meaningful Kubernetes test for that. Made Kubernetes test suite out of the single Kubernetes test in the most simple and effective way I could find. From my POV, this is enough for this PR's scope and I might introduce some improvements in other PRs - provided others agree.

There could be failures, as I might have broken something. Will address the failures once the tests finish - if there are any.

File: ci/Jenkinsfile-kubernetes [hunk 1]
```diff
@@ -302,7 +302,7 @@ spec:
       terminationGracePeriodSeconds: 10
       containers:
         - name: h2o-k8s
-          image: '${registryId}.dkr.ecr.${awsRegion()}.amazonaws.com/sw_kubernetes_repo/sparkling-water:external-backend-${version}'
```
Comment (mn-mikke) [inline ci/Jenkinsfile-kubernetes:305]: Is this change intended?

File: kubernetes/src/main/scala/ai/h2o/sparkling/DataFrameUploadTest.scala [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import org.apache.spark.sql.SparkSession
+
+/**
+  * When running on Kubernetes, H2O-3 disables servlets on non-leader nodes to avoid using a misconfigured cluster.
+  * Yet there are servlets required to be accessible on each node by Sparkling water - those servlets are used by SW
+  * to upload chunks of data to each node.
+  *
+  * This test verifies the data upload servlets are active while running on Kubernetes.
+  */
+object DataFrameUploadTest {
```
Comment (mn-mikke) [inline kubernetes/src/main/scala/ai/h2o/sparkling/DataFrameUploadTest.scala:32]: nit: Would it be possible to rename the object to `DataFrameConversionTest`? I have upload associated with loading from a storage.

File: kubernetes/src/main/scala/ai/h2o/sparkling/InitTest.scala [hunk 3]
```diff
@@ -42,9 +40,7 @@ object InitTest {
     import ai.h2o.sparkling.ml.algos.H2OXGBoost
     val estimator = new H2OXGBoost().setLabelCol("CAPSULE")
     val model = estimator.fit(trainingDF)
-
     model.transform(testingDF).collect()
-    System.out.println("All work done!")
-    hc.stop(true)
   }
+
```
Comment (mn-mikke) [inline kubernetes/src/main/scala/ai/h2o/sparkling/InitTest.scala:45]: nit: extra line

File: kubernetes/src/main/scala/ai/h2o/sparkling/KubernetesTests.scala [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    * Represents a whole suite of Kubernetes tests. Any test not invoked in this method is NOT ran.
+    *
+    * @param spark An instance of SparkSession
+    * @param hc An instance of H2OContext
+    */
+  def runTests(spark: SparkSession, hc: H2OContext): Unit = {
+    InitTest.run(spark, hc)
+    DataFrameUploadTest.run(spark, hc)
+  }
+
```
Comment (mn-mikke) [inline kubernetes/src/main/scala/ai/h2o/sparkling/KubernetesTests.scala:37]: nit: extra line

**Review by mn-mikke — APPROVED** [review]
Minor comments, otherwise LGTM. Please wait with merging your PR, I will try to fix K8s tests first

File: ci/Jenkinsfile-kubernetes (hunk 1 above)
Comment (Pscheidl) [reply ci/Jenkinsfile-kubernetes]: No ! Thank you for catching it.

File: kubernetes/src/main/scala/ai/h2o/sparkling/DataFrameUploadTest.scala (hunk 2 above)
Comment (Pscheidl) [reply kubernetes/src/main/scala/ai/h2o/sparkling/DataFrameUploadTest.scala]: Done. Thank you. Much more consistent.

File: kubernetes/src/main/scala/ai/h2o/sparkling/InitTest.scala (hunk 3 above)
Comment (Pscheidl) [reply kubernetes/src/main/scala/ai/h2o/sparkling/InitTest.scala]: Removed. Thanks.

File: kubernetes/src/main/scala/ai/h2o/sparkling/KubernetesTests.scala (hunk 4 above)
Comment (Pscheidl) [reply kubernetes/src/main/scala/ai/h2o/sparkling/KubernetesTests.scala]: Removed, thank you.

**Comment by Pscheidl** [comment]
@mn-mikke Roger that, not merging. Watching tests in the meantime. I can help btw.

**Comment by Pscheidl** [comment]
Waiting fro other PRs - data have to be redistributed in order for this test to be meaningful.

**Review by mn-mikke — APPROVED** [review]
LGTM, thanks!

File: kubernetes/src/main/scala/ai/h2o/sparkling/DataFrameConversionTest.scala [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  * This test verifies the data upload servlets are active while running on Kubernetes.
+  */
+object DataFrameConversionTest {
+
+  def run(spark: SparkSession, hc: H2OContext): Unit = {
+    import spark.implicits._
+    spark.sparkContext.addFile(
+      "https://raw.githubusercontent.com/h2oai/sparkling-water/master/examples/smalldata/prostate/prostate.csv")
+    val frame = H2OFrame(new URI("file://" + SparkFiles.get("prostate.csv")))
+    val sparkDF = hc.asSparkFrame(frame).withColumn("CAPSULE", $"CAPSULE" cast "string")
```
Comment (mn-mikke) [inline kubernetes/src/main/scala/ai/h2o/sparkling/DataFrameConversionTest.scala:39]: Can you repartition sparkDF before sending to h2oFrame?
